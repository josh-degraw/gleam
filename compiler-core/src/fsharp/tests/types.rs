use crate::assert_fsharp;

#[test]
fn type_inferrence_produces_correct_type_params() {
    assert_fsharp!(
        r#"
type Action(element) {
  Stop
  Continue(element, fn() -> Action(element))
}
pub opaque type Iterator(element) {
  Iterator(continuation: fn() -> Action(element))
}
pub type Step(element, accumulator) {
  Next(element: element, accumulator: accumulator)
  Done
}
pub fn repeatedly(f: fn() -> element) -> Iterator(element) {
 todo
}
 pub fn unfold(
  from initial: acc,
  with f: fn(acc) -> Step(element, acc),
) -> Iterator(element) {
  todo
}
"#,
    );
}

#[test]
fn sanitizes_type_var() {
    assert_fsharp!(
        r#"
pub fn foo(class: class, member: member) {member}
"#,
    );
}
