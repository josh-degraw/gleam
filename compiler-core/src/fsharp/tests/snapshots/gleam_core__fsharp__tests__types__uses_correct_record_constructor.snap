---
source: compiler-core/src/fsharp/tests/types.rs
expression: "\nimport gleam/iterator.{Next}\ntype Cat {\n  Cat(id: Int)\n}\nfn go(cat: Cat) {\n    Next(cat, Cat(id: cat.id + 1))\n    iterator.Next(cat, Cat(id: cat.id + 1))\n}\n"
---
module rec my.``mod``
let Next = gleam.iterator.Next
type private Cat = { id: int64 }
let private go (cat: Cat) = begin
    Next(cat, ({ id = cat.id + 1L }))
    gleam.iterator.Next(cat, ({ id = cat.id + 1L }))
end
